rpc/swap.go

```diff
@@ -46,37 +46,68 @@ func NewSwapService(
 	}
 }
 
-// PastSwapInfo contains a swap ID and its start and end time.
-type PastSwapInfo struct {
-	ID        types.Hash `json:"id" validate:"required"`
-	StartTime time.Time  `json:"startTime" validate:"required"`
-	EndTime   *time.Time `json:"endTime"`
+// PastSwap represents a past swap returned by swap_getPast.
+type PastSwap struct {
+	ID             types.Hash          `json:"id" validate:"required"`
+	Provided       coins.ProvidesCoin  `json:"provided" validate:"required"`
+	ProvidedAmount *apd.Decimal        `json:"providedAmount" validate:"required"`
+	ExpectedAmount *apd.Decimal        `json:"expectedAmount" validate:"required"`
+	ExchangeRate   *coins.ExchangeRate `json:"exchangeRate" validate:"required"`
+	Status         types.Status        `json:"status" validate:"required"`
+	StartTime      time.Time           `json:"startTime" validate:"required"`
+	EndTime        *time.Time          `json:"endTime"`
+}
+
+// GetPastRequest ...
+type GetPastRequest struct {
+	OfferID types.Hash `json:"offerID"`
```

Comment:
You can put `validate:"required"` on this one to make the check for an all-zero hash automatic.

As a side note, on something I learned recently: It works here, because `Hash` is an array. The validator can check any field that resolves to a built-in type for the uninitialized value, but it can't check to see if a structure type has all of its fields uninitialized (which would potentially be a recursive operation). This is because the validator recurses into structures and looks for annotations on the nested fields. Because both `big.Int` and `apd.Decimal` are implemented as structures (and don't have annotations on their nested fields), we can't use the default validation annotations to see if they are zero. We can check that their pointer types are initialized/required, but we can't do a `dive,required` validation on their pointer types.

Reply:
I want to allow empty hashes though, wouldn't adding required defeat the purpose there?

Reply:
Given that it is an optional parameter, I'd use this declaration:
```
OfferID *types.Hash `json:"offerID,omitempty"`
```
You get several benefits:
* the field becomes self documenting that it is optional
* the JSON on the wire, which people do look at, becomes just `{}`
* checking for nil is arguably cleaner than checking for a zero hash